packages/angular_devkit/build_angular/src/utils/server-rendering/esm-in-memory-loader/loader-hooks.ts

```diff
@@ -47,49 +56,93 @@ export function resolve(
+
+  // Check for `./` and `../` relative specifiers
+  const isRelative =
+    specifier[0] === '.' &&
+    (specifier[1] === '/' || (specifier[1] === '.' && specifier[2] === '/'));
+
+  // Relative specifiers from memory file should be based from the parent memory location
+  if (isRelative) {
+    let specifierUrl;
+    try {
+      specifierUrl = new URL(specifier, context.parentURL);
+    } catch {}
+
+    if (
+      specifierUrl?.pathname &&
+      Object.hasOwn(outputFiles, specifierUrl.href.slice(memoryVirtualRootUrl.length))
+    ) {
       return {
         format: 'module',
         shortCircuit: true,
-        // File URLs need to absolute. In Windows these also need to include the drive.
-        // The `/` will be resolved to the drive letter.
-        url: pathToFileURL('/' + normalizedSpecifier).href,
+        url: specifierUrl.href,
       };
     }
+
+    assert.fail(
+      `In-memory ESM relative file should always exist: '${context.parentURL}' --> '${specifier}'`,
+    );
   }
 
+  // Update the parent URL to allow for module resolution for the workspace.
+  // This handles bare specifiers (npm packages) and absolute paths.
   // Defer to the next hook in the chain, which would be the
   // Node.js default resolve if this is the last user-specified loader.
-  return nextResolve(
-    specifier,
-    isBundleEntryPointOrChunk(context) ? { ...context, parentURL: workspaceRootFile } : context,
-  );
+  return nextResolve(specifier, {
+    ...context,
+    parentURL: new URL('index.js', memoryVirtualRootUrl).href,
+  });
 }
 
 export async function load(url: string, context: { format?: string | null }, nextLoad: Function) {
   const { format } = context;
 
-  // CommonJs modules require no transformations and are not in memory.
-  if (format !== 'commonjs' && isFileProtocol(url)) {
-    const filePath = fileURLToPath(url);
-    // Remove '/' or drive letter for Windows that was added in the above 'resolve'.
-    let source = outputFiles[relative('/', filePath)] ?? TRANSFORMED_FILES[filePath];
+  // Load the file from memory if the URL is based in the virtual root
+  if (url.startsWith(memoryVirtualRootUrl)) {
+    const source = outputFiles[url.slice(memoryVirtualRootUrl.length)];
+    assert(source !== undefined, 'Resolved in-memory ESM file should always exist: ' + url);
+
+    // In-memory files have already been transformer during bundling and can be returned directly
+    return {
+      format,
+      shortCircuit: true,
+      source,
+    };
+  }
 
-    if (source === undefined) {
-      source = TRANSFORMED_FILES[filePath] = Buffer.from(
-        await javascriptTransformer.transformFile(filePath),
-      ).toString('utf-8');
-    }
+  // Only module files potentially require transformation. Angular libraries that would
+  // need linking are ESM only.
+  if (format === 'module' && isFileProtocol(url)) {
+    const filePath = fileURLToPath(url);
+    const source = await javascriptTransformer.transformFile(filePath);
```

Comment:
OOC, should we keep caching this?

Reply:
Each ES module is only ever loaded once so the cache won’t get hit.